src/template/transaction/executor.ts

```diff
@@ -0,0 +1,822 @@
+  ctx: TxTemplateContext,
+  ins: z.infer<typeof CompleteTxInstruction>
+) {
+  ctx.log(`Begin CompleteTxInstruction: ${JSONBigInt.stringify(ins)}`);
+  const token = getVariable<string | undefined>(
+    ins.token,
+    ctx.vars,
+    CompleteTxInstruction.shape.token
+  );
+  const changeAddress =
+    getVariable<string | undefined>(
+      ins.changeAddress,
+      ctx.vars,
+      CompleteTxInstruction.shape.changeAddress
+    ) ?? (await interpreter.getChangeAddress(ctx));
+  const address = getVariable<string | undefined>(
+    ins.address,
+    ctx.vars,
+    CompleteTxInstruction.shape.address
+  );
+  const timelock = getVariable<number | undefined>(
+    ins.timelock,
+    ctx.vars,
+    CompleteTxInstruction.shape.timelock
+  );
+  ctx.log(
+    `changeAddress(${changeAddress}) address(${address}) timelock(${timelock}) token(${token})`
+  );
+
+  const tokensToCheck: string[] = [];
+  if (token) {
+    tokensToCheck.push(token);
+  } else {
+    // Check HTR and all tokens on the transaction
+    tokensToCheck.push(NATIVE_TOKEN_UID);
+    ctx.tokens.forEach(tk => {
+      tokensToCheck.push(tk);
+    });
+  }
+
+  // const script = createOutputScriptFromAddress(address, interpreter.getNetwork());
+  const changeScript = createOutputScriptFromAddress(changeAddress, interpreter.getNetwork());
+
+  for (const tokenUid of tokensToCheck) {
+    ctx.log(`Completing tx for token ${tokenUid}`);
+    // Check balances for token.
+    const balance = ctx.balance.getTokenBalance(tokenUid);
+    const tokenData = ctx.addToken(tokenUid);
+    if (balance.tokens > 0) {
+      const value = balance.tokens;
+      // Surplus of token on the inputs, need to add a change output
+      ctx.log(`Creating a change output for ${value} / ${tokenUid}`);
+      // Add balance to the ctx.balance
+      ctx.balance.addOutput(value, tokenUid);
+
+      // Creates an output with the value of the outstanding balance
+      const output = new Output(value, changeScript, { timelock, tokenData });
+      ctx.addOutput(-1, output);
+    } else if (balance.tokens < 0) {
+      const value = -balance.tokens;
+      ctx.log(`Finding inputs for ${value} / ${tokenUid}`);
+      // Surplus of tokens on the outputs, need to select tokens and add inputs
+      const options: IGetUtxosOptions = { token: tokenUid };
+      if (address) {
+        options.filter_address = address;
+      }
+      const { changeAmount, utxos } = await interpreter.getUtxos(value, options);
+
+      // Add utxos as inputs on the transaction
+      const inputs: Input[] = [];
+      for (const utxo of utxos) {
+        ctx.log(`Found utxo with ${utxo.value} of ${utxo.tokenId}`);
+        ctx.log(`Create input ${utxo.index} / ${utxo.txId}`);
+        inputs.push(new Input(utxo.txId, utxo.index));
+      }
+
+      // Update balance
+      for (const input of inputs) {
+        const origTx = await interpreter.getTx(input.hash);
+        ctx.balance.addInput(origTx, input.index);
+      }
```

Comment:
Why don't you do this in the same loop above?

Reply:
Done